torchao/prototype/low_bit_optim/adam.py

```diff
@@ -59,7 +64,7 @@ def _new_buffer(self, p: Tensor, signed: bool):
         local_p = p.to_local() if isinstance(p, DTensor) else p
 
         # follow bitsandbytes, only quantize tensors >= 4096 values
-        if local_p.numel() >= 4096 and local_p.numel() % self.block_size == 0:
+        if p.numel() >= 4096 and p.numel() % self.block_size == 0 and id(p) not in self.exclude_low_bit_optim_params_ids:
```

Comment:
You should keep using `local_p` here for FSDP to work correctly (the check on divisibility should be done on local tensor, not the full tensor)

Furthermore, the check `id(p) not in self.exclude_low_bit_optim_params_ids` should be done before this. Just short-circuit it e.g. (`if p in self.exclude_low_bit_optim_params: return torch.zeros_like(p)`)